cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operations/OperationRunsStore.java

```diff
@@ -0,0 +1,384 @@
+      throws OperationRunNotFoundException, IOException {
+
+    AtomicReference<ScanOperationRunsRequest> requestRef = new AtomicReference<>(request);
+    AtomicReference<String> lastKey = new AtomicReference<>();
+    AtomicInteger currentLimit = new AtomicInteger(request.getLimit());
+
+    while (currentLimit.get() > 0) {
+      AtomicInteger count = new AtomicInteger();
+
+      scanRunsInternal(requestRef.get(), entry -> {
+        lastKey.set(entry.getKey());
+        currentLimit.decrementAndGet();
+        consumer.accept(entry.getKey(), entry.getValue());
+        return count.incrementAndGet() < txBatchSize && currentLimit.get() > 0;
+      });
+
+      if (lastKey.get() == null) {
+        break;
+      }
+
+      ScanOperationRunsRequest nextBatchRequest = ScanOperationRunsRequest
+          .builder(requestRef.get())
+          .setScanFrom(lastKey.get())
+          .setLimit(currentLimit.get())
+          .build();
+      requestRef.set(nextBatchRequest);
+      lastKey.set(null);
+    }
+    return currentLimit.get() == 0;
+  }
+
+
+  private void scanRunsInternal(ScanOperationRunsRequest request,
+      Function<Map.Entry<String, OperationRunDetail<?>>, Boolean> func)
+      throws IOException, OperationRunNotFoundException {
+    Range.Bound startBound = Range.Bound.INCLUSIVE;
+    Range.Bound endBound = Range.Bound.INCLUSIVE;
+    Collection<Field<?>> startFields = new ArrayList<>();
+
+    startFields.add(Fields.stringField(StoreDefinition.OperationRunsStore.NAMESPACE_FIELD,
+        request.getNamespace()));
+
+    Collection<Field<?>> endFields = startFields;
+
+    if (request.getScanFrom() != null) {
+      startBound = Range.Bound.EXCLUSIVE;
+      startFields = getRangeFields(request.getNamespace(), request.getScanFrom());
+    }
+    if (request.getScanTo() != null) {
+      endBound = Range.Bound.EXCLUSIVE;
+      endFields = getRangeFields(request.getNamespace(), request.getScanTo());
+    }
+
+    Range range = Range.create(endFields, endBound, startFields, startBound);
+    Collection<Field<?>> filterIndexes = getFilterIndexes(request.getFilter());
+    int limit = request.getLimit();
+
+    StructuredTable table = getOperationRunsTable(context);
+    try (CloseableIterator<StructuredRow> iterator = table.scan(range, limit,
+        filterIndexes, StoreDefinition.OperationRunsStore.START_TIME_FIELD, SortOrder.DESC)) {
+      boolean keepScanning = true;
+      while (iterator.hasNext() && keepScanning && limit > 0) {
+        StructuredRow row = iterator.next();
+        Map.Entry<String, OperationRunDetail<?>> scanEntry =
+            new SimpleImmutableEntry<String, OperationRunDetail<?>>(
+                row.getString(StoreDefinition.OperationRunsStore.ID_FIELD),
+                GSON.fromJson(row.getString(StoreDefinition.OperationRunsStore.DETAILS_FIELD),
+                    OperationRunDetail.class)
+            );
+        keepScanning = func.apply(scanEntry);
+        limit--;
```

Comment:
limit is already passed to table.scan(), seems like it shouldn't be necessary here.